daffodil-core/src/main/scala/org/apache/daffodil/dsom/ElementBase.scala

```diff
@@ -327,11 +330,11 @@ trait ElementBase
    * nor any PUA-remapped characters. This insures the default/fixed value can still be
    * used for ordinary XML-schema validation outside of Daffodil/DFDL.
    */
-  final lazy val defaultValue: Option[AnyRef] = {
+  final lazy val defaultValue: DataValuePrimitiveOrUseNilForDefaultOrNull = {
```

Comment:
Yeah, so here I think it would be fine, and less confusing, if both DataValue.NoValue and DataValue.UseNilForDefault were of type NoValue, so that defaultValue is of type DataValuePrimitiveOrNoValue, just as DISimple's value slots are. The fact that only defaultValue can hold the sentinel indicating useNilForDefault, .... that's such a hack. Using the type system for this, rather than just letting this thing be NoValue and having a boolean flag, which would be so much more maintainable. 

I truly regret the sin of putting in the UseNilForDefault object hack. Basically because you figured out a way to make it type safe! At the cost of having to understand how that can be possible. We truly should go back to a flag.

Reply:
I have no problem with switching to using just NoValue with a seperate boolean flag.

Short of doing that, we should keep what we are doing explicit.

Keep in mind that UseNilForDefault != NoValue, so all you really accomplish by putting them both in the NoValue type-class is re-introducing the hack of sticking UseNilForDefault where it does not belong.

Reply:
We can keep it the way you have it. I would like explanation of Nullable in scaladoc so that it's less confusing.